packages/common-scripts/src/common.ts

```diff
@@ -1,142 +1,338 @@
-import {
-  isSecp256k1Blake160Address,
-  isSecp256k1Blake160MultisigAddress,
-  prepareSigningEntries as _prepareSigningEntries,
-  isSecp256k1Blake160Script,
-  isSecp256k1Blake160MultisigScript,
-} from "./helper";
 import { getConfig, Config } from "@ckb-lumos/config-manager";
-import lockTimePool from "./locktime_pool";
-import { Address, Header } from "@ckb-lumos/base";
+import locktimePool from "./locktime_pool";
+import {
+  Address,
+  Header,
+  Script,
+  values,
+  Cell,
+  HexString,
+  Hash,
+  PackedSince,
+} from "@ckb-lumos/base";
+import anyoneCanPay from "./anyone_can_pay";
+const { ScriptValue } = values;
+import { Set, List } from "immutable";
+
+/**
+ * CellCollector should be a class which implement CellCollectorInterface.
+ */
+export interface LockScriptInfo {
+  code_hash: Hash;
+  hash_type: "type" | "data";
+  lockScriptInfo: {
+    CellCollector: any;
+    setupInputCell(
+      txSkeleton: TransactionSkeletonType,
+      inputCell: Cell,
+      fromInfo: FromInfo,
+      options: {
+        config?: Config;
+        defaultWitness?: HexString;
+        since?: PackedSince;
+        needCapacity?: HexString;
+      }
+    ): {
+      txSkeleton: TransactionSkeletonType;
+      usedCapacity: HexString;
+    };
+    prepareSigningEntries(
+      txSkeleton: TransactionSkeletonType,
+      options: Options
+    ): TransactionSkeletonType;
+    setupOutputCell?: (
+      txSkeleton: TransactionSkeletonType,
+      outputCell: Cell,
+      options: Options
+    ) => Promise<TransactionSkeletonType>;
+  };
+}
 
+function getPredefinedLockScriptInfos({
+  config = undefined,
+}: Options = {}): LockScriptInfo[] {
+  config = config || getConfig();
+  const secpTemplate = config.SCRIPTS.SECP256K1_BLAKE160!;
+  const multisigTemplate = config.SCRIPTS.SECP256K1_BLAKE160_MULTISIG!;
+  const acpTemplate = config.SCRIPTS.ANYONE_CAN_PAY!;
+
+  return [
+    {
+      code_hash: secpTemplate.CODE_HASH,
+      hash_type: secpTemplate.HASH_TYPE,
+      lockScriptInfo: secp256k1Blake160,
+    },
+    {
+      code_hash: multisigTemplate.CODE_HASH,
+      hash_type: multisigTemplate.HASH_TYPE,
+      lockScriptInfo: secp256k1Blake160Multisig,
+    },
+    {
+      code_hash: acpTemplate.CODE_HASH,
+      hash_type: acpTemplate.HASH_TYPE,
+      lockScriptInfo: anyoneCanPay,
+    },
+  ];
+}
+
+/**
+ *
+ * @param txSkeleton
+ * @param fromInfos
+ * @param toAddress
+ * @param changeAddress
+ * @param amount
+ * @param tipHeader will not use locktime cells if tipHeader not provided
+ * @param options
+ */
 export async function transfer(
   txSkeleton: TransactionSkeletonType,
   fromInfos: FromInfo[],
   toAddress: Address,
   amount: bigint,
-  tipHeader: Header | undefined,
+  changeAddress?: Address,
+  tipHeader?: Header,
   {
-    config,
-    requireToAddress,
-    useLocktimeCellsFirst,
+    config = undefined,
+    useLocktimeCellsFirst = true,
+    customLockScriptInfos = [],
```

Comment:
I was thinking maybe we can keep an array in this module, and provide a `registerCustomLockScriptInfo` method which puts custom lockScriptInfo into this global array. Notice that we are building a framework(we are already injecting current config to a global variable anyway), this way developers do not have to pass-in customLockScriptInfos every time they innvoke `transfer`.

Reply:
In the above design, we also can remove the distinction between predefined lockScriptInfos and customLockScriptsInfos. Predefined ones are just lockScriptInfos that is automatically inserted into the global array.

Reply:
updated.